app/src/main/java/agency/five/codebase/android/movieapp/ui/moviedetails/MovieDetailsViewModel.kt

```diff
@@ -0,0 +1,38 @@
+package agency.five.codebase.android.movieapp.ui.moviedetails
+
+import agency.five.codebase.android.movieapp.data.repository.MovieRepository
+import agency.five.codebase.android.movieapp.ui.favorites.FavoritesViewState
+import agency.five.codebase.android.movieapp.ui.favorites.mapper.FavoritesMapper
+import agency.five.codebase.android.movieapp.ui.moviedetails.mapper.MovieDetailsMapper
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+
+class MovieDetailsViewModel(
+    private val movieRepository: MovieRepository,
+    movieDetailsMapper: MovieDetailsMapper
+) : ViewModel() {
+    private val _movieDetailsViewState = MutableStateFlow(MovieDetailsViewState())
+    val movieDetailsViewState: StateFlow<MovieDetailsViewState> =
+        _movieDetailsViewState.asStateFlow()
+    private val _mapper = movieDetailsMapper
+    init {
+        createViewModel(_mapper)
+    }
+    private fun createViewModel(movieDetailsMapper: MovieDetailsMapper){
+        viewModelScope.launch {
+            movieRepository.movieDetails(_movieDetailsViewState.value.id)
+                .collect { movieDetails ->
+                    _movieDetailsViewState.value =
+                        movieDetailsMapper.toMovieDetailsViewState(movieDetails)
+                }
+        }
+    }
+    fun toggleFavorite(movieId: Int) {
+        movieRepository.toggleFavorite(movieId)
+        createViewModel(_mapper)
+    }
```

Comment:
Let's go through your implementation of this view model to realise what is going on, on how can you improve this code.

First of all, you should provide `movieId` to `MovieDetailsViewModel` through constructor, so it can call `movieRepository.movieDetails(movieId)` with it. [Here](https://insert-koin.io/docs/reference/koin-android/viewmodel#passing-parameters-to-constructor) is how you can provide id with koin. 

In the init function, you are calling `createViewModel` function which launches a coroutine. In this coroutine, you collect  `movieDetails` and publish this value to `movieDetailsViewState`. 
When user clicks on `toggleFavorite`, you are marking movie as favorite through repository, and again calls this createViewModel function to collect new state. 

This is not reactive. Reactive means that you start observing state from repository, and you react to changes emitted from the repository (by producing another view state). By calling `movieRepository.toggleFavorite(movieId)`, you are changing state in the repository. 

I will give you implementation of this view model, so you can better understand how this should look. Also, apply the same pattern to other view models. If you have any concerns, or if something is not clear, please let me know so we can arrange consultations. 

```kotlin
class MovieDetailsViewModel(
    movieId: Int,
    private val movieRepository: MovieRepository,
    movieDetailsMapper: MovieDetailsMapper,
) : ViewModel() {

    val movieDetailsViewState: StateFlow<MovieDetailsViewState> =
        movieRepository.movieDetails(movieId)
            .map(movieDetailsMapper::toMovieDetailsViewState)
            .stateIn(
                scope = viewModelScope,
                started = SharingStarted.WhileSubscribed(5_000),
                initialValue = MovieDetailsViewState.EMPTY,
            )

    fun toggleFavorite(movieId: Int) {
        viewModelScope.launch {
            movieRepository.toggleFavorite(movieId)
        }
    }
}
```